doc/firmware_design_guidelines.md

```diff
@@ -0,0 +1,186 @@
+*A living document of best practices for writing firmware.*
+
+*For coding style guidelines, see the [Zephyr Style Guide](https://docs.zephyrproject.org/latest/contribute/style/index.html).*
+
+## Memory Allocation
+
+The **malloc()** function is not permitted to be used and any dynamic memory allocation should be avoided after initialization. It is incredibly easy to mishandle memory and free routines, and they are unpredictable as to how much time they take to execute. Therefore, they are not safe for safety-critical applications.
+
+All arrays must be statically declared using a fixed number of elements.
```

Comment:
Does this imply that `int array[] = {0, 1, 2};` is not permitted, and that you must instead do `int array[3] = {0, 1, 2};`?